internal/apiendpoint/api_endpoint.go

```diff
@@ -0,0 +1,168 @@
+// Package apiendpoint provides a lightweight API framework for use with River
+// UI. It lets API endpoints be defined, then mounted into an http.ServeMux.
+package apiendpoint
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"time"
+
+	"github.com/riverqueue/riverui/internal/apierror"
+)
+
+// Endpoint is a struct that should be embedded on an API endpoint, and which
+// provides a partial implementation for EndpointInterface.
+type Endpoint[TReq any, TResp any] struct {
+	// Logger used to log information about endpoint execution.
+	logger *slog.Logger
+
+	// Metadata about the endpoint. This is not available until SetMeta is
+	// invoked on the endpoint, which is usually done in Mount.
+	meta *EndpointMeta
+}
+
+func (e *Endpoint[TReq, TResp]) SetLogger(logger *slog.Logger) { e.logger = logger }
+func (e *Endpoint[TReq, TResp]) SetMeta(meta *EndpointMeta)    { e.meta = meta }
+
+// EndpointInterface is an interface to an API endpoint. Some of it is
+// implemented by an embedded Endpoint struct, and some of it should be
+// implemented by the endpoint itself.
+type EndpointInterface[TReq any, TResp any] interface {
+	// Execute executes the API endpoint.
+	//
+	// This should be implemented by each specific API endpoint.
+	Execute(ctx context.Context, req *TReq) (*TResp, error)
+
+	// Meta returns metadata about an API endpoint, like the path it should be
+	// mounted at, and the status code it returns on success.
+	//
+	// This should be implemented by each specific API endpoint.
+	Meta() *EndpointMeta
+
+	// SetLogger sets a logger on the endpoint.
+	//
+	// Implementation inherited from an embedded Endpoint struct.
+	SetLogger(logger *slog.Logger)
+
+	// SetMeta sets metadata on an Endpoint struct after its extracted from a
+	// call to an endpoint's Meta function.
+	//
+	// Implementation inherited from an embedded Endpoint struct.
+	SetMeta(meta *EndpointMeta)
+}
+
+// EndpointMeta is metadata about an API endpoint.
+type EndpointMeta struct {
+	// Path is the API endpoint's path where it should be mounted, which is
+	// passed to http.ServeMux by Mount. It should start with a verb like `GET`
+	// or `POST`, and may contain Go 1.22 path variables like `{name}`, whose
+	// values should be extracted by an endpoint request struct's custom
+	// ExtractRaw implementation.
+	Path string
```

Comment:
From the `Path` name, I definitely didn't expect that this would take a compound string of `METHOD /path`, though it makes sense because that's what `ServeMux.HandleFunc()` takes. However they use the term `pattern` which I think is more appropriate than "path". What do you think about renaming to something like that or another term?

Reply:
Hmm, TBH "pattern" sounds a little weird to me (sounds like a regex?), but you're right, it is what Go's using now, so I guess it's okay. Changed to `Pattern`.